lib/absinthe/type/built_ins/scalars.ex

```diff
@@ -3,23 +3,43 @@ defmodule Absinthe.Type.BuiltIns.Scalars do
 
   @moduledoc false
 
-  scalar :integer, name: "Int" do
-    description """
-    The `Int` scalar type represents non-fractional signed whole numeric values.
-    Int can represent values between `-(2^53 - 1)` and `2^53 - 1` since it is
-    represented in JSON as double-precision floating point numbers specified
-    by [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point).
+  if Application.get_env(:absinthe, :use_legacy_non_compliant_int_scalar_type, true) do
```

Comment:
In my view the flag is backwards here. We can't make a change like this in a point release that might break people's working code. The default needs to be the non compliant int, and then you can opt into the compliant int. In a 2.0 release we can swap this.

Reply:
@benwilson512 

> The default needs to be the non compliant int

The default is the non compliant Int.

---

Maybe the naming is not good? The idea is that the lib will **use the legacy non compliant int scalar type** by default, thus `true` is the default.

You can also see this explained in the documentation I added.

In the `Installation`:

> As explained in Absinthe.Type.Scalar, for legacy reasons, absinthe's default `Int` scalar type is not compliant with the GraphQl spec, thus it is recommended to configure the following for new installations:
> 
> ```elixir
> config :absinthe, use_legacy_non_compliant_int_scalar_type: :false
> ```

<details>
  <summary>How it looks in the doc</summary>
  ![image](https://user-images.githubusercontent.com/5692603/145037756-ea0a4f41-5044-449d-9801-59ae2ba6c8e6.png)
</details>


In the `Absinthe.Type.Scalar`:

> `:integer` - Represents non-fractional signed whole numeric value. **By default it is not** compliant with the GraphQl Specification**, it can represent values between `-(2^53 - 1)` and `2^53 - 1` as specified by [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point). To use the [GraphQl compliant Int](https://spec.graphql.org/October2021/#sec-Int), representing values between `-2^31` and `2^31 - 1`, set `config :absinthe, use_legacy_non_compliant_int_scalar_type: :false` in your config. Note that Absinthe uses the full word `:integer` to identify this type, but its `name` (used by variables, for instance), is `"Int"`.

<details>
  <summary>How it looks in the doc</summary>
  ![Screen Shot 2021-12-07 at 13 27 58](https://user-images.githubusercontent.com/5692603/145038613-dac71300-546c-4e44-bcd4-e81e62fd6ee3.png)
</details>

`In the CHANGELOG`:

> Bug Fix: Add **optional** fix for non compliant built-in scalar Int type. Set `config :absinthe, use_legacy_non_compliant_int_scalar_type: :false` in your config to use the fixed Int type. It is also advisable to upgrade for custom types if you are leveraging the use of integers outside the standard. More information in the [pull request #1131](https://github.com/absinthe-graphql/absinthe/pull/1131).

Reply:
I think he is going for something more along the lines of
```elixir
config :absinthe, use_spec_compliant_int_scalar: true
```
to enable

Reply:
Which would then make this
```elixir
if Application.get_env(:absinthe, :use_spec_compliant_int_scalar, false) do
```

Reply:
This setting is now applied globally, across different schema's in the same application. 

Would it not be possible to apply it per schema by manipulating the scalar in schema phases? This way each schema can individually opt in to these changes.

Reply:
@kdawgwilk Thank you, I see, reversing the name will do the trick then.

Reply:
> Would it not be possible to apply it per schema by manipulating the scalar in schema phases?

@maartenvanvliet making it configurable across schemas would be cool. I'm not sure how would we do this nor what are the implications.
But I also think this is a nice to have that can be left for another pull request, if needed. 
Making this feature would most likely force us to fiddle with more components within the project, and we would be building something to mutate the standard built in types - that should never be done.

Users can always create their custom types if they want the legacy one. The only difference for consumers is the name of the scalar during introspection. 

What are your thoughts?

Reply:
There's a `absinthe/test/absinthe/schema/hydrate_builtins_test.exs` test already that demonstrates how to override a builtin scalar. This relies on the hydrate functionality. 

A similar approach could be used to override the default int scalar. In line with the rest of Absinthe it would an extensible way to override builtin scalars. Arguably the building blocks are there already, and we'd need to find a way to expose it to support the legacy->compliant integer conversion.

Reply:
@maartenvanvliet interesting, sorry I didn't know about the hydrate feature.

Right now you can override the Int with hydrate regardless of the configuration, it just works like it worked before, no existing functionality was changed.
The only things to note is that `serialize_int` (in `Absinthe.Type.BuiltIns.Scalars`) changes depending on the setting. Would having distinct `serialize_ieee_int` and `serialize_graphql_int` help if users want to use them separately?

Software cannot fulfill everyone's dreams and corner cases. Is the feature requested above something nice to have or it is it actually a blocker for this fix? This PR is an optional, opt-in fix, that does not affect any user that does not explicitly says they want it, keeping a all the existing functionality.
If this feature is going forwards, in this pull request, I will probably need more help.

Reply:
I'd be in favor of making it schema specific. Absinthe hardly relies on global configuration (except for some logging I think) and this gives the most control to the user.

Here is a rough version https://gist.github.com/maartenvanvliet/cc3005fdb5950059edc5e0e753bd7c0c where the default integer type is overriden with the spec compliant integer. Needs some cleaning up though.

This can be done per schema by setting the `@pipeline_modifier Absinthe.Phase.Schema.SpecCompliantInt`, so it's not necessary to rely on the global configuration and it remains opt in for now. 

It could be made more ux friendly by e.g. `use Absinthe.Schema, use_spec_compliant_int_scalar: true`, which sets the pipeline modifier for the user.

Reply:
Thank you for the detailed suggestion and the sample code.
I will take a look as submit the necessary changes once I am available.